### ISSUE TITLE
VideoPlayer DirectX - DllNotFound MediaFoundation (missing mf.dll)

### ISSUE THREAD

**Issue by afrodynamics** [MonoGame/MonoGame#5933]
We've had several complaints from users that their application cannot find mf.dll. Our game uses the DirectX backend. We've found that our users can install this on Windows 10 to fix the issue:

https://support.microsoft.com/en-us/help/4016817/media-feature-pack-for-n-edition-of-windows-10-version-1703-april-2017

Generally our players crash when trying to play our intro cinematic in the VideoPlayer. Is it possible for us to ship with this dependency, or for it to be included in standard MG DX builds?

(NOTE: If this is just something we have to require our users to install on their system, then this may not be a bug at all).

**Comment by tomspilman** [comment]
> We've found that our users can install this on Windows 10 to fix the issue:

Yeah... so some systems might never had this dependency installed.  So some crash and others just work.


> be included in standard MG DX builds?

I don't think we can do that.  My guess is that this mf.dll has to be installed into the OS with a bunch of registry junk to work.  We likely cannot just add this DLL into the C# project at build time.

So i don't think we can do that specifically.

However we could possibly protect against the crash and throw something reasonable that the game code can handle and/or maybe show an appropriate error message.

In fact you can likely do this yourself now... catch the exception that occurs in your game code and pop up a dialog leading them to where to download this dependency.

Another approach for MG is to drop the Windows OS video player for our own which we can redistribute.  I forget if this has been discussed before.

Still at the very least the MG docs should be updated to alert developers that this is a dependency for using the Video Player on Windows.

> Is it possible for us to ship with this dependency, 

I am unsure.  Because it seems like there isn't just one dependency to ship here and it may need different ones depending on OS version:

https://support.utomik.com/hc/en-us/articles/115003677505-A-game-asks-me-to-install-Media-Feature-Pack-WMVCore-dll

It suggest the better thing is to redirect users to your own landing page that points to the correct installer to run.

**Comment by afrodynamics** [comment]
> Still at the very least the MG docs should be updated to alert developers that this is a dependency for using the Video Player on Windows.

I would be fully okay with closing this issue and amending the documentation to include mention of this.

I was a little afraid that it would be the sort of thing that is required to be installed on their system. That makes sense, given what the user-facing solution is.

I'm mainly logging this issue here to aid the google searches of future developers.

**Comment by mrhelmut** [comment]
I have a pending documentation update that is mentionning this: #4430

Because the license of Windows Media Player doesn't allow to ship this dependency with other software, the best you can do is to detect the presence of the media feature pack through registery keys and prompt the player to download it manually if it isn't installed: https://github.com/MonoGame/MonoGame/pull/4430/files#diff-8dc49b1ae8de8e9bbca27d3fe4787a20R10

**Comment by tomspilman** [comment]
I keep wondering if we're better off supporting some other open source video player library, but i know other AAA engines still use platform specific video players by default.

I think the best bet is to document things and throw a good exception in the `VideoPlayer` constructor if we cannot initialize the video playback system on the platform.  This allows games to capture the exception then continue without video and/or forward players to an appropriate support page with details on how to address the issue.

**Comment by alexandervrs** [comment]
Isn't any C# solution for FFMPEG? I think that'd work for most platforms
